### PR TITLE
feat(tests): add benchmark for transfers to non-existent receivers

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
## 🗒️ Description
Adds a benchmark that touches many non-existent accounts

## 🔗 Related Issues or PRs

N/A.

## ✅ Checklist

- [x] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx tox -e static
    ```
- [x] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.
- [ ] All: Considered adding an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [ ] All: Set appropriate labels for the changes (only maintainers can apply labels).
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://eest.ethereum.org/main/tests/) are correctly formatted.
- [ ] Tests: For PRs implementing a missed test case, update the [post-mortem document](/ethereum/execution-spec-tests/blob/main/docs/writing_tests/post_mortems.md) to add an entry the list.
- [ ] Ported Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) or [tests/static](/ethereum/execution-spec-tests/blob/main/tests/static) have been assigned `@ported_from` marker.

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: tests/benchmark/compute/scenario/test_transaction_types.py [hunk 1]
```diff
@@ -67,6 +67,14 @@ def get_distinct_receiver_list(
         yield pre.fund_eoa(balance)
 
 
+def get_distinct_unfunded_receiver_list(
+    pre: Alloc,
+) -> Generator[Address, None, None]:
+    """Get a list of distinct unfunded receiver accounts."""
+    while True:
+        yield pre.empty_account()
```
Comment (jochem-brouwer) [inline tests/benchmark/compute/scenario/test_transaction_types.py]: Should be checked if this account does not end up in genesis as empty ([EIP-161 "empty" ](https://eips.ethereum.org/EIPS/eip-161))

**Review by jochem-brouwer — COMMENTED** [review]
This test touches and creates (value sent is 1) many accounts via a transaction. Might also want to expand on touching those (but not creating them -> sending 0 value) and doing this in a transaction by creating them via a CALL loop.

**Comment by MariusVanDerWijden** [comment]
@jochem-brouwer I think the parameterization of the benchmark ("balance", [0,1]) already does the second part (touching but not creating them)
I pushed a second test for the the first part (creating from within a contract)

**Comment by jochem-brouwer** [comment]
You are right, I'm not sure why I missed the part where it also "sends" 0 :sweat_smile:

**Comment by MariusVanDerWijden** [comment]
Hmm I don't get why this fails:
```
 E   execution_testing.base_types.composite_types.Account.BalanceMismatchError: unexpected balance for account 0x57d5192d3682925609bb7f6188a5dc9daeff09ff: want 0x02, got 0x01
 ```

**Comment by jochem-brouwer** [comment]
For the other tests, where the target account(s) exist, the account is initialized with a balance. The test checks if the expected balance is met for the targets. However, in this test we create empty accounts, so only balance = 0 is valid. Balance = 1 will make the test think it yields accounts (which also get 1 more balance) which now have balance of 2. This is incorrect. For the empty test case, skip the test with balance=1:

```python
    if case_id == "a_to_nonexistent" and balance == 1:
        pytest.skip("balance=1 is invalid for a_to_nonexistent")
```

(Should be added to `test_block_full_of_ether_transfers`)

**Comment by jochem-brouwer** [comment]
Actually I think the test for one account to many non-existent already exists: this is the case for balance = 0. It targets these non-existent accounts. So the case `a_to_diff_acc` with `balance=0` is the same as the test created in `test_transaction_types`. So `a_to_nonexistent` does not add a new test but is a duplicate.

The `balance` used here is the balance where the accounts are initialized with (the receivers). It is not the value sent (this is always 1 in the `test_block_full_of_ether_transfers` test). The balances are parametrized as [0, 1]. The zero case is the non-existent account case.

File: tests/benchmark/compute/scenario/test_transaction_types.py [hunk 2]
```diff
@@ -151,6 +165,7 @@ def test_block_full_of_ether_transfers(
     - diff_acc_to_b: multiple senders → one receiver
     - a_to_diff_acc: one sender → multiple receivers
     - diff_acc_to_diff_acc: multiple senders → multiple receivers
+    - a_to_nonexistent: one sender → non-existent receivers
```
Comment (jochem-brouwer) [inline tests/benchmark/compute/scenario/test_transaction_types.py]: This is the same as the `test_block_full_of_ether_transfers[fork_Osaka-benchmark-gas-value_1M-blockchain_test-balance_0-case_id_a_to_diff_acc]` test 😄 👍

File: tests/benchmark/compute/instruction/test_system.py [hunk 3]
```diff
@@ -247,6 +247,42 @@ def test_xcall(
...
+def test_contract_calling_many_addresses(
+    benchmark_test: BenchmarkTestFiller,
+    pre: Alloc,
+    fork: Fork,
+) -> None:
+    """
+    Benchmark a contract that calls many addresses.
+
+    Creates a contract that generates addresses dynamically using KECCAK256
+    and calls each one with 1 wei. The contract loops until it runs out of gas.
```
Comment (jochem-brouwer) [inline tests/benchmark/compute/instruction/test_system.py]: Is the OOG intended? This rollbacks all created accounts, so the state root calculation is rather easy.

File: tests/benchmark/compute/instruction/test_system.py [hunk 4]
```diff
@@ -247,6 +247,42 @@ def test_xcall(
...
+    # Build contract code that generates addresses and calls them
+    # Pattern:
+    # 1. Store NUMBER in memory[0] (32 bytes)
+    # 2. JUMPDEST (loop start)
+    # 3. Update memory[0] with KECCAK256(memory[0], 32)
+    # 4. CALL address from memory[0] (first 20 bytes) with 1 wei
+    # 5. JUMP back to JUMPDEST
+    setup = Op.MSTORE(0, Op.NUMBER())  # Initialize with block number
+    attack_block = (
+        Op.MSTORE(0, Op.SHA3(0, 32))  # Generate new address via hash
```
Comment (jochem-brouwer) [inline tests/benchmark/compute/instruction/test_system.py]: Can also hash once to pick a starting point and then increment the address by 1 each loop. This saves gas by not spending it on the hashing. (Can use the `1` status from the CALL to directly feed in the addition)

Comment (jochem-brouwer) [reply tests/benchmark/compute/instruction/test_system.py]: Note: this targets the same addresses if the block consists of multiple transactions (these are split up due to Fusaka's 7825 gas limit). So if two transactions are in the block, these will both OOG and will both attempt to create the same accounts (if the gas limit of both txs is equal)

**Review by jochem-brouwer — COMMENTED** [review]
Some more comments 😄 👍

**Comment by MariusVanDerWijden** [comment]
> The zero case is the non-existent account case.
I thought that as well, but I somehow convinced myself that its not. I don't remember the specifics, but iirc the other tests have the accounts in the state already, so they are not non-existent, they are existent with 0 balance

**Comment by jochem-brouwer** [comment]
If I check this test: `test_block_full_of_ether_transfers[fork_Osaka-benchmark-gas-value_1M-blockchain_test-balance_0-case_id_a_to_diff_acc]` fixture output then I see a pre state which is almost empty and a post state which has a lot of new accounts. I might also be on some outdated version of EELS though :thinking:

**Comment by MariusVanDerWijden** [comment]
Ah okay, yeah so I will remove this one

File: tests/benchmark/compute/instruction/test_system.py [hunk 5]
```diff
@@ -276,6 +285,8 @@ def test_contract_calling_many_addresses(
             ),
         ),
     )
+    attack_block += Op.JUMPI(Op.GT(Op.GAS, loop_cost), destination)
+    attack_block += Op.RETURN(0, 0)
 
     benchmark_test(
         code_generator=JumpLoopGenerator(
```
Comment (MariusVanDerWijden) [inline tests/benchmark/compute/instruction/test_system.py]: This is not really correct, we don't need a JumpLoopGenerator here anymore, since I added the loop earlier. I am thinking about how to make it cleaner

Comment (MariusVanDerWijden) [reply tests/benchmark/compute/instruction/test_system.py]: I wonder if it makes sense to have a similar generator to `JumpLoopGenerator` that basically does what I'm doing, have an if condition in the loop that stops on some condition.
Basically:
```
JumpLoopNoOOGGenerator(
    setup=setup,
    attack_block=attack_block,
    condition=Op.GT(Op.GAS, loop_cost)
    teardown=Op.Return(0,0)
)
```
Which would do something like 
```
setup + jumpdest + attack_block + condition + jumpi (condition, jumpdest) + teardown
```

wdyt @jochem-brouwer @LouisTsai-Csie

Comment (jochem-brouwer) [reply tests/benchmark/compute/instruction/test_system.py]: @LouisTsai-Csie I think an exit condition for the loop and teardown is useful for the situations which we don't want to run OOG in the JumpLoopGenerator.

Without this feature you can also implement a While loop, as seen here:

https://github.com/ethereum/execution-specs/blob/3a85f3db8fd88c22f8889db6b03e2997a788ed16/tests/benchmark/compute/instruction/test_system.py#L590-L596

Comment (LouisTsai-Csie) [reply tests/benchmark/compute/instruction/test_system.py]: For this scenario, I think we do not need an additional benchmark code generator. The contract logic itself already handles the OOG condition, so we can simply pass a single transaction to the benchmark test wrapper like this:

```python
benchmark_test(
  tx=Transaction(
      to=pre.deploy_contract(code=code, balance=10**8),
      sender=pre.fund_eoa(),
  )
)
```

The wrapper will then create a set of transactions that follows eip7825, and the execution will naturally stop once the transaction runs out of gas. But we need to be really careful with the gas usage here.

https://github.com/ethereum/execution-specs/blob/dca59f62f16089e6a3256979b9d218976570140f/packages/testing/src/execution_testing/specs/benchmark.py#L322-L350

File: tests/benchmark/compute/instruction/test_system.py [hunk 6]
```diff
@@ -247,6 +247,55 @@ def test_xcall(
...
+        + gas_costs.G_LOW * 3  # ~Gluing opcodes
+    )
+    # Build contract code that generates addresses and calls them
+    # Pattern:
+    # 1. Store Keccak256(NUMBER) in memory[0] (32 bytes)
+    # 2. JUMPDEST (loop start)
+    # 4. CALL address from memory[0] (first 20 bytes) with 1 wei
+    # 5. Update memory[0] with ADD(memory[0], 1)
+    # 6. JUMP back to JUMPDEST
+    setup = Op.MSTORE(0, Op.NUMBER())  # Initialize with block number
```
Comment (LouisTsai-Csie) [inline tests/benchmark/compute/instruction/test_system.py]: As mentioned in this comment, if we are running a block with multiple transactions, each transactions will start by the same index, thus sending ether to the same address, which is not our intended logic.

File: tests/benchmark/compute/instruction/test_system.py [hunk 7]
```diff
@@ -247,6 +247,55 @@ def test_xcall(
...
+    destination = Op.JUMPDEST()
+    attack_block = Op.MSTORE(
+        0,
+        Op.ADD(
+            1,
+            Op.CALL(
+                address=Op.MLOAD(0),  # Load address from memory
+                value=1,  # Send 1 wei
+            ),
+        ),
```
Comment (LouisTsai-Csie) [inline tests/benchmark/compute/instruction/test_system.py]: This would send ether to the same addr every time, since the `CALL` result is either 0 or 1. The result of `ADD` would therefore be either (1 + 0) or (1 + 1), meaning that memory slot 0 would always be the same value.

Instead of the relying on the memory layout, i propose using a stack-based solution instead. More details in the following comment.

File: tests/benchmark/compute/instruction/test_system.py [hunk 8]
```diff
@@ -247,6 +247,55 @@ def test_xcall(
...
+    )
+    attack_block += Op.JUMPI(Op.GT(Op.GAS, loop_cost), destination)
```
Comment (LouisTsai-Csie) [inline tests/benchmark/compute/instruction/test_system.py]:
```suggestion
    attack_block += Op.JUMPI(Op.GT(Op.GAS, loop_cost), len(setup))
```

The program count provided should be an index instead of an opcode.

**Review by LouisTsai-Csie — CHANGES_REQUESTED** [review]
I would suggest a stack-based solution for this scenario, this is my idea:

```python
code = (
  Op.SELFBALANCE # [BALANCE]
  + Op.JUMPDEST # [BALANCE]
  + Op.CALL(address=Op.DUP6, value=1) # [1, BALANCE] Here assume every CALL is successful
  + Op.SWAP1 # [BALANCE, 1] 
  + Op.SUB # [BALANCE-1]
  + Op.JUMPI(Op.GT(Op.GAS, loop_cost), len(Op.SELFBALANCE))
)
```

We could pick a reasonable initial balance as the starting address. Since the balance is decreasing every time, we could ensure the target address differs at each call. Even there are multiple transaction, this logic still prevents duplicate ether transfer target.

To prevent OOG issue, we need to make sure this loop cost is calculated precisely. For my implementation, the loop cost would be:

```python
loop_cost = (
    gas_costs.G_JUMPDEST  # 1
    + gas_costs.G_VERY_LOW * 4  # 4× PUSH1 for CALL zero args = 12
    + gas_costs.G_BASE * 2  # 2× GAS = 4
    + gas_costs.G_VERY_LOW * 5  # PUSH1 value, DUP6, SWAP1, SUB, GT = 15
    + gas_costs.G_VERY_LOW * 2  # PUSH1 dest, PUSH2 loop_cost = 6
    + gas_costs.G_HIGH  # JUMPI = 10
    + gas_costs.G_COLD_ACCOUNT_ACCESS  # 2600
    + gas_costs.G_CALL_VALUE  # 9000
    + gas_costs.G_NEW_ACCOUNT  # 25000
    - gas_costs.G_CALL_STIPEND  # -2300 (refunded when callee has no code)
)

```

Please correct me if this is wrong.

---

I think we could even replace every address calculation with a simple `SELFBALANCE` like this:

```python
Op.JUMPDEST
+ Op.POP(Op.CALL(address=Op.SELFBALANCE, value=1))
+ Op.JUMPI(Op.GT(Op.GAS, loop_cost), len(Op.SELFBALANCE))
```

**Comment by LouisTsai-Csie** [comment]
One question - what happens if OOG? Does receiver still get ether or everything is reverted?

**Comment by MariusVanDerWijden** [comment]
We can use the stack instead of the memory. I don't think we should base it on balance though, because then it wouldn't be replay-able twice on bloatnet

File: tests/benchmark/compute/instruction/test_system.py [hunk 9]
```diff
@@ -267,6 +267,59 @@ def _deploy_max_contract_factory(
...
+        + Op.CALL(address=Op.DUP6, value=1)  # [1, BLOCKHASH]
+        + Op.SWAP1  # [BLOCKHASH, 1] // Assume every CALL is successful
+        + Op.SUB  # [BLOCKHASH-1]
+        + Op.JUMPI(Op.GT(Op.GAS, loop_cost), len(setup))
+        + Op.RETURN(0, 0)
+    )
+
+    benchmark_test(
+        tx=Transaction(
+            to=pre.deploy_contract(code=code),
```
Comment (LouisTsai-Csie) [inline tests/benchmark/compute/instruction/test_system.py]:
```suggestion
            to=pre.deploy_contract(code=code, balance=10**18),
```

We should fund this contract, so it could transfer ether to accounts.

File: tests/benchmark/compute/instruction/test_system.py [hunk 10]
```diff
@@ -267,6 +267,59 @@ def _deploy_max_contract_factory(
...
+    benchmark_test: BenchmarkTestFiller,
+    pre: Alloc,
+    fork: Fork,
+) -> None:
+    """
+    Benchmark a contract that calls many addresses.
+
+    Creates a contract that generates addresses dynamically using KECCAK256
+    and calls each one with 1 wei.
+    Stop when the gas is less than the loop cost.
```
Comment (LouisTsai-Csie) [inline tests/benchmark/compute/instruction/test_system.py]: We are using BLOCKHASH(0) to count the address now, not using KECCAK256.

**Review by LouisTsai-Csie — CHANGES_REQUESTED** [review]
left a few small changes. Also, maybe we could clean up some of the comments—I think the code itself is already very clear.

**Review by LouisTsai-Csie — APPROVED** [review]
Add some fixes

File: tests/benchmark/compute/instruction/test_system.py [hunk 11]
```diff
@@ -37,6 +39,63 @@
 )
 
 
+@pytest.mark.parametrize("transfer_amount", [0, 1])
+@pytest.mark.parametrize("opcode", [Op.CALL, Op.CALLCODE])
```
Comment (jochem-brouwer) [inline tests/benchmark/compute/instruction/test_system.py:43]: Why no `DELEGATECALL` and `STATICCALL`? For the value transfers, skip those tests 👍

File: tests/benchmark/compute/instruction/test_system.py [hunk 12]
```diff
@@ -37,6 +39,63 @@
...
+        gas_costs = fork.gas_costs()
+        intrinsic_cost_calc = fork.transaction_intrinsic_cost_calculator()
+
+        iterations = (tx_gas_limit - intrinsic_cost_calc()) // (
+            gas_costs.G_ACCESS_LIST_ADDRESS + cost
+        )
+
+        access_list = [
+            AccessList(
+                address=Address(warm_start_addr - i),
```
Comment (jochem-brouwer) [inline tests/benchmark/compute/instruction/test_system.py]: Adding items to the access list will increase the tx cost and will therefore impact the iterations. Due to Fusaka the transactions will now be split up in multiple ones, but since the setup always pushes the same address, every transaction will interact with the same addresses (which is not what is intended to test here since we want as many unique targets as possible)

File: tests/benchmark/compute/instruction/test_system.py (hunk 11 above)
Comment (spencer-tb) [reply tests/benchmark/compute/instruction/test_system.py:43]: Agree here, happy with the skips. I think we should clean up all the skips in a follow up PR though. We are skipping 1/3 of all benchmark tests that are collected now lol

**Review by spencer-tb — APPROVED** [review]
